apex_launchtest/apex_launchtest/apex_runner.py

```diff
@@ -24,79 +24,98 @@
-    # (launch_description, test_context).  This wrapper function normalizes things
-    # so we always get a tuple, sometimes with an empty dictionary for the test_context
-    def wrapper(*args, **kwargs):
-        result = launch_description_fn(*args, **kwargs)
-        if isinstance(result, tuple):
-            return result
-        else:
-            return result, {}
+class _LaunchDiedException(Exception):
+    pass
 
-    return wrapper
 
-
-class ApexRunner(object):
+class _RunnerWorker():
 
     def __init__(self,
-                 gen_launch_description_fn,
-                 test_module,
+                 test_run,
                  launch_file_arguments=[],
                  debug=False):
-        """
-        Create an ApexRunner object.
-
-        :param callable gen_launch_description_fn: A function that returns a ros2 LaunchDesription
-        for launching the processes under test.  This function should take a callable as a
-        parameter which will be called when the processes under test are ready for the test to
-        start
-        """
-        self._gen_launch_description_fn = gen_launch_description_fn
-        self._test_module = test_module
+        self._test_run = test_run
         self._launch_service = LaunchService(debug=debug)
         self._processes_launched = threading.Event()  # To signal when all processes started
         self._tests_completed = threading.Event()  # To signal when all the tests have finished
         self._launch_file_arguments = launch_file_arguments
 
         # Can't run LaunchService.run on another thread :-(
         # See https://github.com/ros2/launch/issues/126
-        # Instead, we'll let the tests run on another thread
+        #
+        # It would be simpler if we could run the pre-shutdown test and the post-shutdown tests on
+        # one thread, and run the launch on another thead.
+        #
+        # Instead, we'll run the pre-shutdown tests on a background thread concurrent with the
+        # launch on the main thread.  Once the launch is stopped, we'll run the post-shutdown
+        # tests on the main thread
         self._test_tr = threading.Thread(
             target=self._run_test,
             name='test_runner_thread',
             daemon=True
         )
 
-    def get_launch_description(self):
-        return _normalize_ld(self._gen_launch_description_fn)(lambda: None)[0]
-
     def run(self):
         """
         Launch the processes under test and run the tests.
 
         :return: A tuple of two unittest.Results - one for tests that ran while processes were
         active, and another set for tests that ran after processes were shutdown
         """
-        test_ld, test_context = _normalize_ld(
-            self._gen_launch_description_fn
-        )(lambda: self._processes_launched.set())
-
-        # Data to squirrel away for post-shutdown tests
-        self.proc_info = ActiveProcInfoHandler()
```

Comment:
Since we're doing 100% of the loading and binding before any tests run now, we can reduce the scope of the proc_into and proc_output helper objects.

Previously, we delayed loading the tests until we were ready to run them, but the new `TestRun` object over in `loader.py` makes it easier to do the loading and binding and parametrization up front